### PR TITLE
feat: nested fast path refactor 

### PR DESCRIPTION
**Pull request by elbuo8**
## Description

Refactors nested TTU and Userset implementations. The overall number of queries produced continues to be the same as the original implementation. Continues to load the first level of usersets along side the terminal type to intersect on before progressing to resolving recursion. 

The main change is preventing N*M traversal. This PR introduces a Breadth First approach bounding the concurrency of the traversal while reducing the overall cost of consumption of intermediate responses and reducing overall contention of the `sync` objects. The consumer of intermediate responses follows the exact flow we have in `consumeDispatches` and `consumeUsersets`, tho lining them up to be refactored into a single method across all flows.

It also fixes a data race introduced in the refactor of `combinedIterator` in a previous PR which started being exhibited in this PR 😅 

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

If you haven't done so yet, we would appreciate it if you could star the [OpenFGA repository](https://github.com/openfga/openfga). :)

### CODE REVIEW

File: internal/graph/tuplemapper.go [hunk 1]
```diff
@@ -89,12 +91,44 @@ func (n NestedTTUMapper) doMap(t *openfgav1.TupleKey) (string, error) {
 	return t.GetUser(), nil
 }
 
+type ObjectIDMapper struct {
```
Comment (adriantam) [inline internal/graph/tuplemapper.go:94]: How is this different from NestedTTUMapper?  Would it make sense to just use NestedTTUMapper?  If the name makes it confusing, perhaps rename NestedTTUMapper to be more generic such as tupleKeyMapper?

Comment (elbuo8) [reply internal/graph/tuplemapper.go:94]: One returns the `User`, the other returns `Object`

Comment (adriantam) [reply internal/graph/tuplemapper.go:94]: Ahhh...got it.

File: internal/graph/check.go [hunk 2]
```diff
@@ -979,227 +982,170 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
...
+// group:1#member@group:3#member
+// group:2#member@group:a#member
+// group:3#member@group:a#member
+// Note that both group:2#member and group:3#member has group:a#member. However, they are not cycles.
+
+func (c *LocalChecker) breadthFirstNestedMatch(ctx context.Context, req *ResolveCheckRequest, mapping *nestedMapping, visitedUserset *sync.Map, currentUsersetLevel []string, usersetFromUser storage.SortedSet, checkOutcomeChan chan checkOutcome) {
+	req.GetRequestMetadata().Depth--
+	if req.GetRequestMetadata().Depth == 0 {
+		concurrency.TrySendThroughChannel(ctx, checkOutcome{err: ErrResolutionDepthExceeded}, checkOutcomeChan)
+		close(checkOutcomeChan)
```
[the remark this answers is not in the document]
Comment (elbuo8) [reply internal/graph/check.go:1003]: There is no concurrency in this function. It must finish all goroutines before scheduling a recursive all once again.

File: pkg/storage/iterators.go [hunk 3]
```diff
@@ -53,28 +55,36 @@ func (c *combinedIterator[T]) Next(ctx context.Context) (T, error) {
...
 	}
 
+	c.mu.Lock()
 	iter := c.pending[0]
 	val, err := iter.Next(ctx)
 	if err != nil {
 		if errors.Is(err, ErrIteratorDone) {
 			c.pending = c.pending[1:]
 			c.done = append(c.done, iter)
+			c.mu.Unlock()
```
Comment (elbuo8) [inline pkg/storage/iterators.go:65]: The introduction of this mutex is necessary as `c.pending` and `c.done` get accessed via multiple goroutines and can cause data race issues (and there are other methods which ultimately access these properties such as `.Stop`, thus a mutex is required.

**Review by adriantam — APPROVED** [review]

**Pull request merged**